tests/Main.hs

```diff
@@ -111,6 +110,38 @@ spec = do
           ]
         , "result" .= Null
         ]
+    it "ValidationFailed (2)" $ do
+      let r :: FormResult LoginFields Text Text
+          r = ValidationFailed (e0 <> e1)
+          e0 = mkFieldError (pick @"username" :| [pick @"password"]) msg
+          e1 = mkFieldError (pick @"username" :| [pick @"remember_me"]) msg
+          msg = "Something" :: Text
+      toJSON r `shouldBe` object
+        [ "parse_error" .= Null
+        , "field_errors" .= object
+          [ "username" .= object
+            [ "password" .= ("Something" :: Text)
+            , "remember_me" .= ("Something" :: Text)
+            ]
+          ]
+        , "result" .= Null
+        ]
+    it "ValidationFailed (3)" $ do
+      let r :: FormResult LoginFields Text Text
+          r = ValidationFailed (e0 <> e1)
+          e0 = mkFieldError (nes $ pick @"username") msg
+          e1 = mkFieldError (pick @"username" :| [pick @"remember_me"]) msg
+          msg = "Something" :: Text
+      toJSON r `shouldBe` object
+        [ "parse_error" .= Null
+        , "field_errors" .= object
+          [ "username" .= object
+            -- the validation error at deeper level `e1` has precedence
+            [ "remember_me" .= ("Something" :: Text)
```

Comment:
Hmm, *precedence*. I'm not sure this format is good now. It looks like the current option with dot-separated fields allows to represent any errors without needing precedence rules:

```
{
  "field_errors": {
    "username": "Foo",
    "username.remember_me": "Bar",
  }
}
```

Reply:
I was a bit thrown by this too, but I think the precedence is justified here because if the validation fails on a nested field then you don't need to validate its parent - it's implied that it will fail.

Reply:
Well, of the ideas of the library is that we want to collect all validation errors "in parallel". So basically although we do this in our code just fine, this is a bit spoiled during conversion to JSON.

I'm sorry that I asked you to prepare the PR and you spent your time, but I think we should choose the representation that preserves all validation errors always as the default.

On a flip side though, you now have `unFieldError` so you can do your own conversion and this code you would need to write anyway for it.

Reply:
That's all good, I appreciate the level of consideration you put into this and I agree that if this doesn't work for everyone then it's not a good default option.

I'm still going to use this format for my purposes, but I'd be happy to contribute it here as an alternative option for error formatting. I'll update the PR accordingly.